### DISCUSSION TITLE
Dynamic Redirects, best way?

### DISCUSSION THREAD

**Discussion by ndimatteo** [vercel/next.js#15344] [Help] [answered]
Hey there!

**I'm wondering what everyone's opinions are on how to properly generate page redirects that are defined outside of Next?**

For example, I have data coming from Sanity.io where a client would maintain a list of redirects (with `from`, `to`, and `isPermanent` fields). How can I take that data and generate redirects within NextJS? It's possible that the redirect `from` URL could cross multiple dynamic and static pages defined within Next.

My initial solution was to write these redirects to `vercel.json` in my build scripts, but unfortunately, Vercel reads the config file BEFORE build scripts are run so the redirects are never read through.

I feel this solution would be such a clean solution for setting these up, but with how Vercel deploys projects it seems unlikely this solution will work without a core update from the Vercel team.

**The goal is to avoid:**
1. hardcoding the redirects in the codebase.
2. Setting up a 3rd-party service to write these on deploy
3. Adding these into each page template via `getStaticProps` (especially when the redirects are "unknown")

Thoughts?

**Comment by ArrayKnight** [comment]
I was just about to start testing this exact same approach, using a headless CMS on a separate domain from the site to manage redirect and then a pre-build step to update the vercel.json redirects via webhook and local script.

We had questions about whether the vercel.json could even be modified (possibly not being allowed to write to fs at all) or if we could even manage to do it early enough in the process for vercel to pick up the new version rather than the repo version.

Glad I found this before investing a bunch of time down this path.

But we had another path to venture:
Using our webhook to trigger a Github action, which modifies and commits the vercel.json file. Not quite as simple or clean but it bypasses vercel completely.

Did you come up with a different/better solution?

**Comment by ndimatteo** [comment]
@ArrayKnight since posting this there have been a number of updates to the Next platform and I was able to easily do this with the new way they handle redirects. [You can read more about it in their docs](https://nextjs.org/docs/api-reference/next.config.js/redirects), but I found it as simple as this for a Sanity powered site:

```javascript
// get redirects from Sanity for Vercel
async function fetchSanityRedirects() {
  const data = await client.fetch(
    `*[_type == "redirect"]{ from, to, isPermanent }`
  )

  const redirects = data.map((redirect) => {
    return {
      source: `/${redirect.from}`,
      destination: `/${redirect.to}`,
      permanent: redirect.isPermanent,
    }
  })

  return redirects
}

module.exports = {
  async redirects() {
    const sanityRedirects = await fetchSanityRedirects()
    return sanityRedirects
  }
}
```

**Reply by ArrayKnight** [reply]
I appreciate the assist!

**Reply by thiniulian** [reply]
I'm doing this as well, but I have a hard time answering 2 questions:
1. Should I redeploy my app when a redirect is added in the 3rd party platform?
2. When is the redirects function called and how many times? - I'm thinking of distributed deployments, how can I be sure all of them are using the same set of redirects?


In the past I've done it in different way, creating a prebuild npm script that creates a json file which gets used where we need it, but the  questions remain valid.

**Reply by spencersmb** [reply]
I'm looking at implementing this feature as well, and my first thought is yes - a redeployment will be needed I think because next config is not available to the browser and is used by the next server(vercel) at build time only. I'm not 100% on that but it seems like it from reading here: https://nextjs.org/docs/api-reference/next.config.js/introduction

@thiniulian or @ArrayKnight - have you made any progress on this front? Did you end up implementing the redirects from your cms?

I have a bunch of redirects in a plugin called pretty-links on my wordpress site that I'm trying to move to headless with next.js

Cheers,
Spencer

**Reply by ArrayKnight** [reply]
Sorry, I haven't been on that project in quite some time so my specific recollection may not be perfect. But if I am remembering correctly, yea, a build was necessary, but because the site was already using the CMS webhook to trigger a build, and updating the redirects within the CMS triggered this (just like any other content update), it all worked as intended

**Reply by thiniulian** [reply]
@spencersmb I'm currently working at this. As @ArrayKnight mentioned the rebuild should happen and the best option is with the webhook trigger.
In my case, I use NX to manage multiple similar next.js projects and it's complicating the situation even further since I can't import/require things in next.config.js as I would want. So I'm looking at the option to generate the next.config.js from a custom script before the build since I'll do the rebuild anyway.

**Reply by rodrigo-arias** [reply]
I'm doing something similar with WordPress and [Redirection](https://wordpress.org/plugins/redirection/). The two problems I have faced:
- Need to rebuild to get new redirects added
- Validation to prevent the build from breaking if invalid redirects are added in the CMS. On this point, the hard part is that the regexp don't match directly with the format of [Next.js Redirects](https://nextjs.org/docs/api-reference/next.config.js/redirects)

**Reply by danijelglisic** [reply]
Hey, I have error in next.config.js file  because client is not recognized but when I import it there is error because import statement outside a module. How I can solve this like you in this example?

**Reply by ndimatteo** [reply]
@danijelglisic you'll need to import the Sanity client in `next.config.js` like this to avoid that error:

```js
const sanityClient = require('@sanity/client')
```

**Reply by notflip** [reply]
@rodrigo-arias Which method are you using now? (I know it's 2 years later, but I'm curious if you ever got around these two problems you mention

**Reply by rodrigo-arias** [reply]
@notflip Not really. I tried two approaches:
- A custom nextConfig wrapper that fetched, validated, and normalized redirects and added feedback if something was off, along with a cron job that ran a couple of times daily and triggered a rebuild if new redirects were added. It worked but was far from ideal.
- The official Middleware approach. It solved the rebuild issue but required keeping Redis in sync since users didn't want to leave the CMS to manage redirects. This gets around the limit of 1,024 redirects but adds too much complexity.
﻿

**Comment by Kyoss79** [comment]
To be honest, this is such a bummer and kind of a dealbreaker for our company.
It's really sad that it took us this long to discover this behaviour in nextjs.

Just a short headsup for everyone who stumbles over this issue just like me and is clueless.

NextJS does only generate the redirects from next.config.js at **build-time**. Although it is an async function,
that function is only called during the build.

I was under the impression, that this function would get called repeatedly in production, and was wondering how often that would happen. The answer is **zero** times.

So you can make redirects kind of dynamic, as in load them from an external service, but you can not make them dynamic, as in they will update automatically in production.

Triggering a new build for a new redirect per webhook is kind of a no-go for big sites who basically have hundreds or thousand of redirects. (don't ask me why, for a typo, for a rename of the url, for SEO purposes). This would lead to dozens of redundant and concurrent builds every week in my current project.

What I really needed was an ISR-like method that gets called every XX seconds and repupulates the redirects.

**Reply by rodrigo-arias** [reply]
Faced the same challenges sometime ago when I researched this  https://github.com/vercel/next.js/discussions/15344#discussioncomment-2554070

**Reply by jaandrews** [reply]
I ran into this as well. Ended up having to add middleware to get the redirects via an api call instead of setting the redirects in the config so that new redirects could be added without rebuilding the nextjs side. Definitely a performance hit, since it now has to hit the server on each request, but unavoidable given how redirects behave right now. Below is the middleware.ts file I used handle redirects.

```tsx
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

export default async function middleware(request: NextRequest) {
  var currentDomain = request.headers.get('x-ms-original-url'); //this is used to get the actual url in azure because the host header isn't the actual url.
  if (!currentDomain) {
    currentDomain = 'https://' + request.headers.get('host');
  }
  //passing along the whole domain into the api call, since my build supports multiple domains and some redirects are domain specific.
  var response = await fetch(`https://www.cms.com/redirect/?url=${encodeURIComponent(new URL(request.nextUrl.pathname, currentDomain).href + request.nextUrl.search)}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json'
    }
  });
  if (response.status === 200) {
    var redirect = (await response.json());
    if (redirect && redirect.destination) {
      return NextResponse.redirect(new URL(redirect.destination), {
        status: redirect.permanent ? 308 : 307
      });
    }
  }
}
 
// See "Matching Paths" below to learn more
export const config = {
  matcher: [
    '/((?!api|_next/static|_next/image|favicon.ico).*)'
  ]
};
```

**Reply by thiniulian** [reply]
Have you tried caching the redirects with Vercel KV? Should be a smaller performance hit, but with a higher cost.

**Reply by jaandrews** [reply]
The site this is for has to run on azure, so I'm not able to use vercel hosting features. Others can certainly do so though. There would still be the performance hit for the first load though.

**Reply by smakman** [reply]
When using Vercel you could store your redirects in Vercel's Edge Config and handle the redirects in your middleware:

```js
// middleware.js
import { NextResponse } from "next/server";
import { get } from "@vercel/edge-config";

export async function middleware(request) {
  const redirect = (await get("redirects")).find(({ source }) => source === request.nextUrl.pathname);

  if (redirect) {
    return NextResponse.redirect(new URL(redirect.destination, request.url), {
      status: redirect.permanent ? 308 : 307,
    });
  }
}

export const config = {
  matcher: [
    '/((?!api|_next/static|_next/image|favicon.ico).*)'
  ]
};
```

Your Vercel Edge Config can be updated with the [Vercel REST API](https://vercel.com/docs/storage/edge-config/vercel-api) and would look like:

```js
{
  "redirects": [
    {
      "source": "/from",
      "destination": "/to",
      "permanent": true
    }
  ]
}
```

**Reply by Kyoss79** [reply]
I mean sure, this works. But is it feasible? If hundreds of your users request the same URL, its going to make hundreds of middleware requests to your Edge Config repeatedly. Even if there are zero results for a given URL.
I am not happy with implementing "solutions" like that.

**Reply by smakman** [reply]
If you want dynamic redirects you somehow have to match the current route against an external source. It's mentioned [here](https://vercel.com/docs/storage/edge-config) in the Vercel docs as a possible use case for their Edge Config. It also mentions that the latency is negligible (0-15ms). Egde Config does have its [limitations](https://vercel.com/docs/storage/edge-config/edge-config-limits) though, mainly in terms of number of reads and store size.

**Reply by Kyoss79** [reply]
@smakman sure, I do agree. But if you have several "negligible" calls on your page that starts to sum up. As I said, those are 99% of the time totally unneeded calls anyway.

My current solution would be to put my project behind a provider that let's me configure redirects even before hitting the NextJS Site. Like Cloudflare. 

Another solution no one has mentioned yet, but may be interesting for some: You can create redirects in getStaticProps. So instead of slowing down every request with a middleware, you can ask your backend (or Edge config) in getStaticProps if this page should be redirected.

**Reply by smakman** [reply]
getStaticProps will only run when a page gets statically (re)rendered, not on every request.

**Comment by leerob** [comment, marked as answer]
Please check out our new docs here!

https://nextjs.org/docs/app/building-your-application/routing/redirecting

**Reply by rodrigo-arias** [reply]
@leerob Would you consider adding a kind of On-Demand ISR API that allows us to revalidate the redirects defined in the `next.config.js` without having to rebuild the entire site?
I understand that an alternative solution can be built using the middleware, but that has some performance costs, and mitigating them requires more complex solutions and probably Redis or an equivalent in Vercel.
Thanks!

**Reply by michaelroth-xx** [reply]
would be interested in this as well! a revalidate possibility in next.config would be awesome

**Reply by davit2001** [reply]
If you want to do permenant redirect for a dynamic page where redirect url depends on a api response in that case you should create a Next.js api route like this

```
export async function POST(request: Request) {
	const { url } = await request.json();
	return NextResponse.redirect(url, 301);
}
```

After doing this it still not redirect to the new page it will just show in a logs that page is redirected with 301 status

To solve that issue you should use `router.replace('your_url')` or `router.refresh()` method to actually redirect the page

I know that this is a weird solution but it is the only solution which I found because Next.js `redirect()` redirect page with 307 status
and there is no way to redirect page with 301 or other custom status for a dynamic page where redirect url depends on a api response
